common/src/main/java/org/conscrypt/DESEDESecretKeyFactory.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.conscrypt;
+
+import java.security.InvalidKeyException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactorySpi;
+import javax.crypto.spec.DESedeKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * An implementation of {@link javax.crypto.SecretKeyFactory} for use with DESEDE keys.  This
+ * class supports {@link SecretKeySpec} and {@link DESedeKeySpec} for key specs.
+ *
+ * @hide
+ */
+@Internal
+public class DESEDESecretKeyFactory extends SecretKeyFactorySpi {
+    @Override
+    protected SecretKey engineGenerateSecret(KeySpec keySpec) throws InvalidKeySpecException {
+        if (keySpec == null) {
+            throw new InvalidKeySpecException("Null KeySpec");
+        }
+        if (keySpec instanceof SecretKeySpec) {
+            return (SecretKey) keySpec;
```

Comment:
It seems like this should be checking partiy of the key. DESede is a little bit unique in that it only uses 112-bit or 168-bit but the key can also be supplied as 128 or 192-bits of data. I think `DESedeKeySpec` even has an [isParityAdjusted()](https://docs.oracle.com/javase/7/docs/api/javax/crypto/spec/DESedeKeySpec.html#isParityAdjusted%28byte[],%20int%29) method which might help?

I guess this could be relied upon failing later if it's the wrong size when it is actually used.